rfcs/text/0428-type-inheritance-and-forking.md

```diff
@@ -0,0 +1,950 @@
+
+The duplication happened on the expanded type resulting in combining/expanding `Person` and `Employee`, but it would also have been possible to expand the entire type, getting rid of `Being`. This would have resulted in the following type hierarchy:
+
+```txt
+      name◄────┐
+               │
+    tenure◄────MyEmployee
+               │
+occupation◄────┘
+```
+
+And like the previous example, the `MyEmployee` type would not be a subtype of `Person` anymore. It also wouldn't be a subtype of `Being` either - the reusability of the types is lost in this case. Ideally, type duplication would be able to preserve as many supertype relations as possible.
+
+## Addressing previous considerations
+
+In the [Type System RFC](./0352-graph-type-system.md#interfacing-with-types-1) the following statement was made:
+
+> The update messages of both the current and new systems make use of partial schemas, merging the schema given in the message contents with the existing Entity Type. This may or may not be the desired semantics of updating, and could lead to undesired behavior. In that case, the semantics can be changed to treat updates as a complete replacement. This is to be decided and can be considered out of scope for this RFC as it touches on inheritance/forking concepts.
+
+As the new Block Protocol type system doesn't require a delta-based storage approach for its schemas, it's unclear what the advantages of partial schemas for update requests are. We propose that we should treat type updates as complete replacements, so implementation is much more straightforward for embedding applications. Partial schema updates also add some level of indirection, and that may obfuscate error sources and error reasons. Therefore, updates to types must be complete replacements rather than partial schema updates.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+## Defining extended entity types
+
+In the Block Protocol, we will allow type extension through the `allOf` JSON Schema keyword which specifies an array of schemas that will have to validate together.
+
+We'll add the following fields to the existing Entity Type meta schema definition:
+
+```json
+{
+  "$id": "https://blockprotocol.org/type-system/0.3/schema/meta/entity-type",
+  "type": "object",
+  ...,
+  "properties": {
+    ...,
+    "allOf": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "$ref": {
+            "$comment": "Valid reference to an existing Entity Type version",
+            "$ref": "$ref": "https://blockprotocol.org/types/modules/graph/0.3/schema/versioned-url"
```

Comment:
```suggestion
            "$ref": "https://blockprotocol.org/types/modules/graph/0.3/schema/versioned-url"
```

Reply:
Thanks! My vim bindings likely screwed that one up..

Reply:
No worries 😄